### ISSUE TITLE
Add deployment id

### ISSUE THREAD

**Issue by nearmax** [nearprotocol/nearcore#2403]
As suggested by @k06a it would be convenient to have deployment id stored with each account. Specifically, every time we deploy a new contract we should update `deployment_id` stored in the contract.

Implementations to consider:
* Storing deployment id as a nonce that increases every time the contract is deployed.
  * Pros: Easy versioning -- it is easy to have code that only works with the contract if its version is between X and Y nonces;
  * Cons: We cannot prove that deployment id was generated at certain block index/hash, unless we separately store `touched_at_block` field for the account;
* Storing deployment id as a hash of block id and the contract blob.

**Comment by nearmax** [comment]
Adding @evgenykuzyakov 's comment:

> It would be great to add an option to a function call action to be able to enforce contract_hash .
> I think we can do this in non-breaking fashion. Add a new action, that is just a check and a no-op if the check passes, but fails if the check fails.
> Which can be done post-mainnet

**Comment by vgrichina** [comment]
yeah, not clear why have separate `deployment_id` given we already can check `code_hash`. Basically there just needs to be new version of `FunctionCall` action which enforces given `code_hash`.

**Comment by k06a** [comment]
I hope we can define it similar to CREATE2:
```
contract.deployment_id = hash(deployer.deployment_id, contract.code_hash, salt)
```

**Comment by olonho** [comment]
Indeed ability to conditionally invoke  other contract iff its hash matches expectation + convenient mechanism of storing and updating expectation hashes looks like a nice feature.

**Comment by vgrichina** [comment]
@k06a why is salt needed? Should it be just `hash(contract.address, contract.code_hash)`?

**Comment by bowenwang1996** [comment]
Transferring to NEPs since it requires design discussion
